server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -305,10 +309,73 @@ public void onFailure(Exception e) {
         }
     }
 
-    private void failDecommissionedNodes(ClusterState state) {
+    private void drainNodesWithDecommissionedAttribute(ClusterState state, DecommissionRequest decommissionRequest) {
+        decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.DRAINING, new ActionListener<>() {
+            @Override
+            public void onResponse(DecommissionStatus status) {
+                logger.info("updated the decommission status to [{}]", status);
+                // set the weights
+                checkWeightsAndScheduleDecommission(decommissionRequest);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(
+                    () -> new ParameterizedMessage(
+                        "failed to update decommission status for attribute [{}] to [{}]",
+                        decommissionRequest.getDecommissionAttribute().toString(),
+                        DecommissionStatus.DRAINING
+                    ),
+                    e
+                );
+                // since we are not able to update the status, we will clear the voting config exclusion we have set earlier
+                clearVotingConfigExclusionAndUpdateStatus(false, false);
+            }
+        });
+    }
+
+    void checkWeightsAndScheduleDecommission(DecommissionRequest decommissionRequest) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        if (decommissionAttributeMetadata == null) {
+            clearVotingConfigExclusionAndUpdateStatus(false, false);
+            return;
+        }
+        assert decommissionAttributeMetadata.status().equals(DecommissionStatus.DRAINING)
+            : "Unexpected status encountered while decommissioning nodes.";
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        ensureDecommissionedAttributeWeightsAreSet(state, decommissionAttribute);
+        // Schedule the node decommission process after the weights are successfully set.
+        scheduleNodesDecommissionOnTimeout(decommissionRequest.getDrainingTimeout());
+    }
+
+    private void ensureDecommissionedAttributeWeightsAreSet(ClusterState state, DecommissionAttribute decommissionAttribute) {
+        WeightedRoutingMetadata weightedRoutingMetadata = state.metadata().weightedRoutingMetadata();
+        assert weightedRoutingMetadata != null && weightedRoutingMetadata.getWeightedRouting() != null;
+        assert weightedRoutingMetadata.getWeightedRouting().weights().get(decommissionAttribute.attributeValue()) == 0.0;
```

Comment:
this won't run in prod environment. Also remove this check entirely. We are handling it as part of #4839

Reply:
Removed.